### PR TITLE
Add TS and epistemic uncertainty examples

### PR DESCRIPTION
**Pull request by vtaquet**
# Description

Add two new examples:
- A time-series example in which MAPIE is used to estimate prediction intervals associated with forecast
- Another example using the x*sin(x) function but this time to capture both uncertainties
 
## Type of change

Please check options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] I have read the [contributing guidelines](https://github.com/simai-ml/MAPIE/blob/master/CONTRIBUTING.rst)
- [x] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files
- [x] Linting passes successfully : `flake8 . --exclude=doc`
- [x] Typing passes successfully : `mypy mapie examples --strict`
- [x] Unit tests pass successfully : `pytest -vs --doctest-modules mapie`
- [x] Coverage is 100% : `pytest -vs --doctest-modules --cov-branch --cov=mapie --pyargs mapie`
- [x] Documentation builds successfully : `cd doc; make clean; make html`

### CODE REVIEW

File: examples/plot_both_uncertainties.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    funct: Any,
+    mu: float,
+    sigma: float,
+    n_samples: int,
+    noise: float
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    """
+    Generate noisy 1D data with normal distribution from given function
+    and noise standard deviation.
+    """
```
Comment (gmartinonQM) [inline examples/plot_both_uncertainties.py:35]: docstring incomplète

File: examples/plot_both_uncertainties.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+
+# Functions for generating our dataset
+def x_sinx(x: np.ndarray) -> Any:
+    """One-dimensional x*sin(x) function."""
+    return x*np.sin(x)
+
+
+def get_1d_data_with_normal_distrib(
+    funct: Any,
```
Comment (gmartinonQM) [inline examples/plot_both_uncertainties.py:26]: Pour les fonctions, utiliser un alias `F` dans le typing. L'alias peut être défini comme suit en haut du module, après les imports

```
from typing import TypeVar, Callable, Any

F = TypeVar("F", bound=Callable[..., Any])

File: examples/plot_both_uncertainties.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    n_samples: int,
+    noise: float
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    """
+    Generate noisy 1D data with normal distribution from given function
+    and noise standard deviation.
+    """
+    np.random.seed(42)
+    X_train = np.random.normal(mu, sigma, n_samples)
+    X_test = np.arange(mu-4*sigma, mu+4*sigma, sigma/20.)
```
Comment (gmartinonQM) [inline examples/plot_both_uncertainties.py:38]: pas très PEP8 ces signes `-` dans espace autour

File: examples/plot_both_uncertainties.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    y_pred: np.ndarray,
+    y_pred_low: np.ndarray,
+    y_pred_up: np.ndarray,
+    ax: plt.Axes,
+    title: str
+) -> None:
+    ax.set_xlabel("x")
+    ax.set_ylabel("y")
+    ax.set_xlim([-10, 10])
+    ax.set_ylim([np.min(y_test)*1.3, np.max(y_test)*1.3])
```
Comment (gmartinonQM) [inline examples/plot_both_uncertainties.py:93]: Pourquoi préciser les `xlim` et `ylim` ? Les valeurs automatiques ne sont pas convenables ?

File: examples/plot_both_uncertainties.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+fig, ax = plt.subplots(1, 1, figsize=(7, 5))
+ax.set_xlim([-8, 8])
+ax.set_ylim([0, 4])
+for strategy in STRATEGIES:
+    ax.plot(X_test, y_pis[strategy][:, 1, 0] - y_pis[strategy][:, 0, 0])
+ax.axhline(1.96*2*noise, ls="--", color="k")
+ax.set_xlabel("x")
+ax.set_ylabel("Prediction Interval Width")
+ax.legend(list(STRATEGIES.keys()) + ["True width"], fontsize=8)
+plt.show()
```
Comment (gmartinonQM) [inline examples/plot_both_uncertainties.py:132]: Si je comprends bien, c'est l'exemple iso à la section "Tutorial" de la doc c'est ça ?

File: examples/plot_timeseries_example.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+The electricity demand features daily and weekly seasonalities and is impacted
+by the temperature, considered here as a exogeneous variable.
+
+The data is modelled by a Random Forest model with a
+:class:`sklearn.model_selection.RandomizedSearchCV` using a sequential
+`sklearn.model_selection.TimeSeriesSplit` cross validation, in which the
+training set is prior to the validation set.
+The best model is then feeded into :class:`mapie.estimators.MapieRegressor`
+to estimate the associated prediction intervals.
+We consider the standard CV+ resampling method.
```
Comment (gmartinonQM) [inline examples/plot_timeseries_example.py:21]: Ajouter un paragraphe "known limitations", qui précise qu'en général, il n'y a plus de garanties théoriques pour le TS, mais que en l'occurrence, ça marche bien empiriquement notamment car il n'y a pas de tendance manifest

File: examples/plot_timeseries_example.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+# Load input data and feature engineering
+demand_df = pd.read_csv(
+    "data/demand_temperature.csv",
+    parse_dates=True,
+    index_col=0
+)
+demand_df["Date"] = pd.to_datetime(demand_df.index)
+demand_df["Weekofyear"] = demand_df.Date.dt.isocalendar().week.astype('int64')
+demand_df["Weekday"] = demand_df.Date.dt.isocalendar().day.astype('int64')
```
Comment (gmartinonQM) [inline examples/plot_timeseries_example.py:39]: Tiens je ne connaissais pas `isocalendar`. Pourquoi pas directement : `.dt.day` ?

File: examples/plot_timeseries_example.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    cv=tscv,
+    scoring="neg_root_mean_squared_error",
+    random_state=random_state,
+    verbose=0,
+    n_jobs=-1,
+)
+cv_obj.fit(X_train, y_train)
+best_est = cv_obj.best_estimator_
+
+# Estimate prediction intervals on test set with best estimator
```
Comment (gmartinonQM) [inline examples/plot_timeseries_example.py:74]: Si possible, renvoyer vers l'exemple nested cv pour plus de détails sur ce point touchy

**Review by gmartinonQM — CHANGES_REQUESTED** [review]

File: examples/plot_both_uncertainties.py (hunk 1 above)
Comment (vtaquet) [reply examples/plot_both_uncertainties.py:35]: Docstring complétée.

File: examples/plot_both_uncertainties.py (hunk 2 above)
Comment (vtaquet) [reply examples/plot_both_uncertainties.py:26]: Alias ajouté.

File: examples/plot_both_uncertainties.py (hunk 3 above)
Comment (vtaquet) [reply examples/plot_both_uncertainties.py:38]: Corrigé.

File: examples/plot_both_uncertainties.py (hunk 4 above)
Comment (vtaquet) [reply examples/plot_both_uncertainties.py:93]: Non car la fonction diverge fortement, rendant la figure illisible.

File: examples/plot_both_uncertainties.py (hunk 5 above)
Comment (vtaquet) [reply examples/plot_both_uncertainties.py:132]: Avec du bruit homoscédastique en plus.

File: examples/plot_timeseries_example.py (hunk 7 above)
Comment (vtaquet) [reply examples/plot_timeseries_example.py:39]: `.dt.day` sera bientôt obsolète d'après la documentation pandas, il faudra utiliser `isocalendar` dans les futures versions de pandas.

File: examples/plot_timeseries_example.py (hunk 8 above)
Comment (vtaquet) [reply examples/plot_timeseries_example.py:74]: Commentaire ajouté.

File: examples/plot_timeseries_example.py (hunk 6 above)
Comment (vtaquet) [reply examples/plot_timeseries_example.py:21]: J'ai ajouté le paragraphe ci-dessous. 

Je ne vois qu'une seule grosse limite à l'utilisation d'un `KFold` classique sur des données TS, le fait qu'on utilise des données d'entraînement ultérieures aux données de validation pour estimer les résidus, ce qui est normalement incompatible avec l'ordre chronologique des données. Cependant, je pense que les garanties mathématiques restent valables avec cet objet, même avec une TS avec tendance. Les données de test deviendraient alors "out-of-distribution" mais l'erreur épistémique augmenterait.   

Comme discuté avec Nicolas, l'utilisation d'un `TimeSeriesSplit` est plus pertinente pour ce type de données mais on casserait en l'occurrence les garanties mathématiques, a fortiori sur des données avec tendance. 

```
We wish to emphasize one main limitation with this example. 
We use a standard cross-validation in Mapie to estimate the prediction intervals,
through the `sklearn.model_selection.KFold()` object. 
Residuals are therefore estimated using models trained on data with higher indices
than the validation data, which is inappropriate for time-series data. 
Howerver, using a `sklearn.model_selection.TimeSeriesSplit` cross validation object for
estimating the residuals breaks the theoretical guarantees of the Jackknife+ and CV+
methods.
```

File: examples/plot_both_uncertainties.py [hunk 9]
```diff
@@ -32,10 +33,33 @@ def get_1d_data_with_normal_distrib(
     """
     Generate noisy 1D data with normal distribution from given function
     and noise standard deviation.
+
+    Parameters
+    ----------
+    funct : F
+        Based function used to generate the dataset.
```
Comment (gmartinonQM) [inline examples/plot_both_uncertainties.py:40]: typo: base

**Review by gmartinonQM — CHANGES_REQUESTED** [review]

File: examples/plot_both_uncertainties.py (hunk 9 above)
Comment (vtaquet) [reply examples/plot_both_uncertainties.py:40]: Corrigée.
